### PR TITLE
Handling SSCS1 document type

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1212

### Change description ###
SSCS1 document type should be converted into 'Form'

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java [hunk 1]
```diff
@@ -126,6 +126,7 @@ private static DocumentType mapDocumentType(InputDocumentType inputDocumentType)
             case COVERSHEET:
                 return DocumentType.COVERSHEET;
             case FORM:
+            case SSCS1:
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java:129]: I think this means we should remove from schema validation if present and require for it to be strictly just `form`

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java:129]: do you mean metafile-schema.json? @doncem

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 2]
```diff
@@ -79,8 +82,19 @@ private void assertSameScannableItems(Envelope dbEnvelope, InputEnvelope zipEnve
...
+            .extracting(ScannableItem::getOcrData)
+            .extracting(this::convertToInputOcrData)
+            .usingRecursiveFieldByFieldElementComparator()
+            .containsAll(
+                zipEnvelope.scannableItems.stream().map(item -> item.ocrData).collect(toList())
+            );
+        assertThat(dbEnvelope.getScannableItems())
+            .extracting(ScannableItem::getDocumentType)
+            .usingDefaultComparator()
+            .containsExactly(CHERISHED, OTHER, FORM, DocumentType.COVERSHEET);
```
Comment (mdayican) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:97]: I think better to write another test just for scannableItems

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java:129]: yes this is the only schema we have. but this change is very serious as it purely depends on a contract among each service. plus we insist to be like that anyway

File: src/integrationTest/resources/envelope.json [hunk 3]
```diff
@@ -20,7 +20,7 @@
           "file_name": "1111002.pdf",
           "notes": "test",
           "document_uuid": "0fa1ab60-f836-43aa-8c65-b07cc9bebcbe",
-          "document_type": "other"
+          "document_type": "form"
```
Comment (lgonczar) [inline src/integrationTest/resources/envelope.json:23]: Why this change?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java (hunk 1 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java:129]: We have to keep accepting SSCS1 for now. We'll remove it once Exela has done their change. Unfortunately, for the time being we have to deal with SSCS1 document types coming in metadata files.

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapper.java:129]: That's what I thought, Thanks for clarifying this @lgonczar

File: src/integrationTest/resources/envelope.json (hunk 3 above)
Comment (apereverzin) [reply src/integrationTest/resources/envelope.json:23]: otherwise EnvelopeControllerTest is failing. @lgonczar

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 2 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:97]: I agree with @mdayican. At first, we didn't do this conversion of document types, so there was just simple comparison to do. We should have added such test in the past when that bit got introduced. Having a separate test focusing on document type and subtype conversion would definitely be an improvement.

File: src/integrationTest/resources/envelope.json (hunk 3 above)
Comment (nitinprabhu) [reply src/integrationTest/resources/envelope.json:23]: test data looks incorrect to me we would never receive lower case `other` or `form` for document type.

Comment (lgonczar) [reply src/integrationTest/resources/envelope.json:23]: I don't see anything invalid or new about this. Why would it be failing?

Comment (lgonczar) [reply src/integrationTest/resources/envelope.json:23]: Didn't see Nitin's comment

Comment (lgonczar) [reply src/integrationTest/resources/envelope.json:23]: I've just checked and this is the expected envelope in DB, as returned by an endpoint. `other` is one of the valid values (see `DocumentType` enum). The expectation changed because input data is SSCS1. All good. I was confused by the file name.

Comment (apereverzin) [reply src/integrationTest/resources/envelope.json:23]: @lgonczar Exactly

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 4]
```diff
@@ -79,10 +84,27 @@ private void assertSameScannableItems(Envelope dbEnvelope, InputEnvelope zipEnve
 
         assertThat(dbEnvelope.getScannableItems())
             .extracting(this::convertToInputScannableItem)
-            .usingRecursiveFieldByFieldElementComparator()
+            .usingElementComparatorIgnoringFields("ocrData", "documentType")
             .containsAll(zipEnvelope.scannableItems);
     }
 
+    private void assertSameOcrData(Envelope dbEnvelope, InputEnvelope zipEnvelope) {
```
Comment (mdayican) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:91]: I meant another test to keep assertions less per test and not affect existing test case,

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 2 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java]: @mdayican @lgonczar Added tests

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 5]
```diff
@@ -34,8 +34,24 @@
...
+
+        assertSamePayments(dbEnvelope, zipEnvelope);
+        assertThat(dbEnvelope.getScannableItems()).isEmpty();
+        assertSameNonScannableItems(dbEnvelope, zipEnvelope);
+
+        assertDbEnvelopeSpecificProperties(container, dbEnvelope);
+    }
+
+    @Test
+    public void should_map_zip_envelope_without_sscs1_scannble_items_correctly() throws Exception {
```
Comment (mdayican) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:53]: is it a typo `scannble`

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 6]
```diff
@@ -44,24 +60,42 @@ public void should_map_zip_envelope_correctly() throws Exception {
...
+        assertSameOcrData(dbEnvelope, zipEnvelope);
+        assertDocumentTypes(dbEnvelope, CHERISHED, OTHER, DocumentType.COVERSHEET);
+        assertSameNonScannableItems(dbEnvelope, zipEnvelope);
+        assertRightScannableItemsHaveWarnings(dbEnvelope, ocrValidationWarnings);
+
+        assertDbEnvelopeSpecificProperties(container, dbEnvelope);
+    }
+
+    @Test
+    public void should_map_zip_envelope_with_sscs1_scannble_item_correctly() throws Exception {
```
Comment (mdayican) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:76]: should_map_zip_envelope_with_scannable_items_without_sscs1_item_correctly ?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 7]
```diff
@@ -17,46 +17,85 @@
...
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.DocumentType.CHERISHED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.DocumentType.FORM;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.DocumentType.OTHER;
 
 public class EnvelopeMapperTest {
 
     @Test
-    public void should_map_zip_envelope_correctly() throws Exception {
-        InputEnvelope zipEnvelope = getEnvelopeFromMetafile();
+    public void should_map_zip_envelope_without_scannable_items_correctly() throws Exception {
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:37]: not sure why we need this test case ? Envelope mapper is generic previous one was correct

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/helper/EnvelopeCreator.java [hunk 8]
```diff
@@ -49,12 +50,12 @@
     private EnvelopeCreator() {
     }
 
-    public static InputStream getMetaFile() {
-        return EnvelopeCreator.class.getResourceAsStream("/metafiles/valid/from-spec.json");
+    private static InputStream getMetaFile(String fileName) {
+        return EnvelopeCreator.class.getResourceAsStream(METAFILE_BASE_PATH + fileName);
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/helper/EnvelopeCreator.java:54]: How about providing a full path to the resource in the call? This would make it a bit clearer what resource we're referencing, as the path provides some information.

File: src/test/resources/metafiles/valid/from-spec-no-scannable-items.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (lgonczar) [inline src/test/resources/metafiles/valid/from-spec-no-scannable-items.json:1]: I'd remove `from-spec-` prefix from these file names, as it doesn't really say anything (and is a bit confusing to me).

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 4 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:116]: I agree - one test for this general verification and a separate test just focusing on those document types (there could be an envelope with all of them and the test would verify if each of them has been properly mapped, ignoring all other fields from the envelope).

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 10]
```diff
@@ -17,46 +17,69 @@
...
+        InputEnvelope zipEnvelope = getEnvelopeFromMetafile(METAFILE_BASE_PATH + "no-scannable-items.json");
+        String container = "container1";
+
+        Envelope dbEnvelope = EnvelopeMapper.toDbEnvelope(zipEnvelope, container, Optional.empty());
+
+        assertSameFields(dbEnvelope, zipEnvelope);
+
+        assertSamePayments(dbEnvelope, zipEnvelope);
+        assertThat(dbEnvelope.getScannableItems()).isEmpty();
+        assertSameNonScannableItems(dbEnvelope, zipEnvelope);
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:48]: We could also keep scannable item checking here (that check doesn't take ocr and document types into account):
```
assertSameScannableItems(dbEnvelope, zipEnvelope);
```

And then the name of the method could stay as it was.

Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:48]: Same with ocr data.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 11]
```diff
@@ -17,46 +17,69 @@
...
         assertSameScannableItems(dbEnvelope, zipEnvelope);
-        assertSameNonScannableItems(dbEnvelope, zipEnvelope);
+        assertSameOcrData(dbEnvelope, zipEnvelope);
+        assertDocumentTypes(dbEnvelope, CHERISHED, OTHER, DocumentType.COVERSHEET);
         assertRightScannableItemsHaveWarnings(dbEnvelope, ocrValidationWarnings);
+    }
 
-        // properties specific to DB envelope
+    @Test
+    public void should_map_zip_envelope_with_scannable_items_with_sscs1_item_correctly() throws Exception {
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:71]: How about:
* naming it `should_map_document_types_and_subtypes_correctly()`
* including all existing document types in the test json
* verifying document types along with subtypes?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 12]
```diff
@@ -79,10 +114,27 @@ private void assertSameScannableItems(Envelope dbEnvelope, InputEnvelope zipEnve
...
+            .containsAll(
+                zipEnvelope.scannableItems.stream().map(item -> item.ocrData).collect(toList())
+            );
+    }
+
+    private void assertDocumentTypes(Envelope dbEnvelope, DocumentType... documentTypes) {
+        assertThat(dbEnvelope.getScannableItems())
+            .extracting(ScannableItem::getDocumentType)
+            .usingDefaultComparator()
+            .containsExactly(documentTypes);
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:135]: We need to make sure order is preserved.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 13]
```diff
@@ -17,46 +17,69 @@
...
+
+        assertSamePayments(dbEnvelope, zipEnvelope);
+        assertThat(dbEnvelope.getScannableItems()).isEmpty();
+        assertSameNonScannableItems(dbEnvelope, zipEnvelope);
+
+        assertDbEnvelopeSpecificProperties(container, dbEnvelope);
+    }
+
+    @Test
+    public void should_map_zip_envelope_with_scannable_items_without_sscs1_item_correctly() throws Exception {
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:54]: I don't think we need this test method if we have a separate test method verifying the mapping of documents.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 12 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:135]: This method containsExactly preserves order unlike containsExactlyInAnyOrder

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 11 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:71]: ok

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 13 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:54]: ok

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 12 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:135]: aaa, true

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 14]
```diff
@@ -41,22 +45,49 @@ public void should_map_zip_envelope_correctly() throws Exception {
...
-        assertThat(dbEnvelope.getCaseNumber()).isEqualTo(zipEnvelope.caseNumber);
-        assertThat(dbEnvelope.getPreviousServiceCaseReference()).isEqualTo(zipEnvelope.previousServiceCaseReference);
-        assertThat(dbEnvelope.getPoBox()).isEqualTo(zipEnvelope.poBox);
-        assertThat(dbEnvelope.getJurisdiction()).isEqualTo(zipEnvelope.jurisdiction);
-        assertThat(dbEnvelope.getDeliveryDate()).isEqualTo(zipEnvelope.deliveryDate);
-        assertThat(dbEnvelope.getOpeningDate()).isEqualTo(zipEnvelope.openingDate);
-        assertThat(dbEnvelope.getZipFileCreateddate()).isEqualTo(zipEnvelope.zipFileCreateddate);
-        assertThat(dbEnvelope.getZipFileName()).isEqualTo(zipEnvelope.zipFileName);
-        assertThat(dbEnvelope.getClassification()).isEqualTo(zipEnvelope.classification);
+        assertSameFields(dbEnvelope, zipEnvelope);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:48]: Please hold back on refactoring so that it's easier for reviewers to identify the actual changes related to the bug/feature in a PR.
Separate PRs are welcome.

Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java]: Done @rkondratowicz

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 15]
```diff
@@ -56,7 +60,42 @@ public void should_map_zip_envelope_correctly() throws Exception {
...
+        assertSameOcrData(dbEnvelope, zipEnvelope);
+        assertDocumentTypes(
+            dbEnvelope,
+            CHERISHED,
+            OTHER,
+            OTHER,
+            OTHER,
+            OTHER,
+            FORM,
+            DocumentType.COVERSHEET,
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:82]: can be static import as other doc types above

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 16]
```diff
@@ -79,10 +118,34 @@ private void assertSameScannableItems(Envelope dbEnvelope, InputEnvelope zipEnve
 
         assertThat(dbEnvelope.getScannableItems())
             .extracting(this::convertToInputScannableItem)
-            .usingRecursiveFieldByFieldElementComparator()
+            .usingElementComparatorIgnoringFields("ocrData", "documentType", "documentSubtype")
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:121]: is it not this already comparing ocr doctype and subtype then do we need this extra test `should_map_document_types_and_subtypes_correctly` ?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java [hunk 17]
```diff
@@ -56,7 +60,42 @@ public void should_map_zip_envelope_correctly() throws Exception {
         assertSameNonScannableItems(dbEnvelope, zipEnvelope);
         assertRightScannableItemsHaveWarnings(dbEnvelope, ocrValidationWarnings);
 
-        // properties specific to DB envelope
+        assertDbEnvelopeSpecificProperties(container, dbEnvelope);
+    }
+
+    @Test
+    public void should_map_document_types_and_subtypes_correctly() throws Exception {
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:67]: method says doc type and subtype but you are also checking ocr data ?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 15 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:82]: there is already static import of DocumentSubtype.COVERSHEET

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 17 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:67]: Have renamed

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java (hunk 16 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/model/mapper/EnvelopeMapperTest.java:121]: Next test is focused on comparing types, subtypes and ocr data
